src/storage-client/src/controller.rs

```diff
@@ -2496,24 +2497,21 @@ where
         }
     }
 
-    /// Closes the identified shards from further reads or writes.
-    #[allow(dead_code)]
-    async fn finalize_shards<I>(&mut self, shards: I)
+    /// Using `(shard, since, upper)`, sets the shard's since to at least the
+    /// provided value if `Some`, and likewise for the shard's upper. Returns a
+    /// `BTreeMap` of the resulting values.
+    ///
+    /// Note that this function operates on persist shards and makes no effort
+    /// to update the in-memory state of the controller. However, you can use
+    /// the return values to propagate the appropriate changes to the
+    /// controller's frontier tracking.
+    async fn set_shard_since_upper<I>(
```

Comment:
I have a vague feeling that this could be dangerous. Anything that appends empty data, just to move the `upper` forward has the potential of telling users something that is not "true". We say we know that there was no data during that time period, while we don't necessarily know it's true.

If this is related to the other commit that advances the `upper` to ensure the `since`/`upper` invariant that we're talking about there, this would heighten my sense of nervousness. :sweat_smile:

Reply:
Reverted all of this.